src/basicfuns.jl

```diff
@@ -312,9 +312,23 @@ Return `log(exp(x) + exp(y))`, avoiding intermediate overflow/undeflow, and hand
 non-finite values.
 """
 function logaddexp(x::Real, y::Real)
-    # ensure Δ = 0 if x = y = ± Inf
-    Δ = x == y ? zero(x - y) : abs(x - y)
-    max(x, y) + log1pexp(-Δ)
+    a, b = map(float, promote(x, y))
+    T = typeof(a)
+
+    if (isnan(a) || isnan(b)) return T(NaN) end
+
+    # Compute `maxval = max(a, b), diff = -abs(a - b)`
+    # in a faster type-stable way
+    if a < b
+        diff = a - b
+        maxval = b
+    else
+        diff = b - a
+        maxval = a
+    end
+
+    # `diff` is NaN if both `a` and `b` are infinite
+    isnan(diff) ? maxval : maxval + log1pexp(diff)
```

Comment:
IMO the following is simpler and in particular avoids any explicit type calculations:
```suggestion
    # Compute max = Base.max(x, y) and diff = x == y ? zero(x - y) : -abs(x - y)
    # in a faster type-stable way
    a, b = promote(x, y)
    if a < b
        diff = a - b
        max = b
    else
        # ensure diff = 0 if a = b = ± Inf
        diff = a == b ? zero(a - b) : b - a
        max = !isnan(b) ? a : b
    end
    return max + log1pexp(diff)
```

On my computer it is also faster than the master branch and the current version of the PR:
```julia
julia> using LogExpFunctions, BenchmarkTools

julia> @btime logaddexp(x, y) setup=(x = 1000 * randn(); y = 1000 * randn()); # master branch
  2.370 ns (0 allocations: 0 bytes)                                    

julia> @btime logaddexp(x, y) setup=(x = 1000 * randn(); y = 1000 * randn()); # current proposal
  1.933 ns (0 allocations: 0 bytes)

julia> @btime logaddexp(x, y) setup=(x = 1000 * randn(); y = 1000 * randn()); # proposal
  1.504 ns (0 allocations: 0 bytes)  
```